src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs

```diff
@@ -146,15 +147,27 @@ public static Diagnostic Filter(Diagnostic d, int warningLevelOption, NullableCo
                 report = isEnabledByDefault ? ReportDiagnostic.Default : ReportDiagnostic.Suppress;
             }
 
+            var position = location.SourceSpan.Start;
+
             bool isNullableFlowAnalysisWarning = ErrorFacts.NullableFlowAnalysisWarnings.Contains(id);
+            if (isNullableFlowAnalysisWarning)
+            {
+                var nullableWarningsGloballyEnabled = nullableOption == NullableContextOptions.Enable || nullableOption == NullableContextOptions.Warnings;
+                var nullableWarningsEnabled = tree?.GetNullableDirectiveState(position).WarningsState ?? nullableWarningsGloballyEnabled;
+                if (!nullableWarningsEnabled)
+                {
+                    return ReportDiagnostic.Suppress;
+                }
+            }
+
             if (report == ReportDiagnostic.Suppress && // check options (/nowarn)
                 !isNullableFlowAnalysisWarning)
```

Comment:
>!isNullableFlowAnalysisWarning) [](start = 16, length = 31)

do we still need this?

Reply:
Looks like it doesn't do anything any more, will remove.

Reply:
Done